src/stream/src/executor/over_window/state/mod.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeSet;
+
+use risingwave_common::types::{Datum, ScalarImpl};
+use risingwave_expr::expr::WindowFuncKind;
+use smallvec::SmallVec;
+
+use super::call::WindowFuncCall;
+use super::MemcmpEncoded;
+
+mod lag;
+mod lead;
+
+/// Unique and ordered identifier for a row in internal states.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub(super) struct StateKey {
+    pub order_key: ScalarImpl,
+    pub encoded_pk: MemcmpEncoded,
+}
+
+#[derive(Debug)]
+pub(super) struct StatePos<'a> {
+    /// Only 2 cases in which the `key` is `None`:
+    /// 1. The state is empty.
+    /// 2. It's a pure preceding window, and all ready outputs are consumed.
+    pub key: Option<&'a StateKey>,
+    pub is_ready: bool,
+}
+
+#[derive(Debug)]
+pub(super) struct StateOutput {
+    /// Window function return value for the current ready window frame.
+    pub return_value: Datum,
+    /// Hint for the executor to evict unneeded rows from the state table.
+    pub evict_hint: StateEvictHint,
+}
+
+#[derive(Debug)]
+pub(super) enum StateEvictHint {
+    /// Use a set instead of a single key to avoid state table iter or too many range delete.
+    /// Shouldn't be empty set.
+    CanEvict(BTreeSet<StateKey>),
+    /// State keys from the specified key are still required, so must be kept in the state table.
+    CannotEvict(StateKey),
+}
+
+impl StateEvictHint {
+    pub fn intersect(self, other: StateEvictHint) -> StateEvictHint {
+        use StateEvictHint::*;
+        match (self, other) {
+            (CanEvict(a), CanEvict(b)) => {
+                // Example:
+                // a = CanEvict({1, 2, 3})
+                // b = CanEvict({2, 3, 4})
+                // a.intersect(b) = CanEvict({1, 2, 3})
+                let a_last = a.last().unwrap();
+                let b_last = b.last().unwrap();
+                let last = std::cmp::min(a_last, b_last).clone();
+                CanEvict(a.into_iter().chain(b).filter(|k| k <= &last).collect())
```

Comment:
What about using `split_off` and `union` here?

Reply:
`split_off` seems not suitable here. I changed to

```rust
                CanEvict(
                    a.into_iter()
                        .take_while(|k| k <= &last)
                        .chain(b.into_iter().take_while(|k| k <= &last))
                        .collect(),
                )
```